### DISCUSSION TITLE
Only first picture presented preloads

### DISCUSSION THREAD

**Discussion by SophieFromm** [jspsych/jsPsych#833] [Q&A]
Hi,

I am using an older jsPsych script from another person and am trying to debug it. The picture preloading does not seem to work because the image elements appear twice in the cache when looking at the developer tools in the browser. It looks like only the first picuture that is presented loads from the cache, because this always exists only once.

![screenshot4github](https://user-images.githubusercontent.com/56911897/86443149-23bd7780-bd0f-11ea-93fb-ddcb96099d7a.PNG)

I have this in my js script where "state_name+'_stim_'+stims[1]+'_deact.png" is just the variable that calls the images:
```

state_name = state_names[state];
state_color = state_colors[state];
var images = ['var images = ['img/earth_planet.png', 'img/earth_stim_1.png', 'img/earth_stim_2.png', ..]
				
				if (stage==-1) { // timeout	at first level
					$('#jspsych-space-daw-bottom-stim-left').css('background-image', 'url(img/'+state_name+'_stim_'+stims[0]+'_deact.png)');
					$('#jspsych-space-daw-bottom-stim-right').css('background-image', 'url(img/'+state_name+'_stim_'+stims[1]+'_deact.png)');
				}
				
				if (stage==1) { // choice stage
					display_element.html('');
					$('.jspsych-display-element').css('background-image', 'url("img/'+state_name+'_planet.png")');				
					
					display_element.append($('<div>', {
						id: 'jspsych-space-daw-top-stim-left',
					}));
....
```

...and this in my html:

```
jsPsych.preloadImages(images, function(){ startExperiment(); });

	function startExperiment(){
		jsPsych.init({
			experiment_structure: experiment,
...
`
```

The pictures are quite large and load really slow. So I wondered whether my picture preloading is really working or if the "loading twice" indicates that it is not presenting the preloaded pictures, but instead loads them again. I would appreciate any help. Thanks!!

**Comment by jodeleeuw** [comment]
Do the developer tools show the images loading at the start of the experiment too? Trying to think what the issue might be. Do you happen to know what version of jsPsych this is?

**Reply by SophieFromm** [reply]
Yes the images are there from when I load the task and as soon as a certain picture is displayed it appears twice. Certainly not jsPsych 6, but I am not exactly sure - it must from 2016 or earlier. Thanks for your effort to help!

**Reply by jodeleeuw** [reply]
While I don't think there was a major change in how preloading was handled, I'm not sure. It may be that more recent versions of jsPsych would handle this better.

You could try the method [suggested here](https://github.com/jspsych/jsPsych/discussions/724) to split up the preloading and load images closer to when they are needed.

**Comment by SophieFromm** [comment]
Hi again,

I tried the suggested method, but it did not solve my problem. I also tried to preload images by loading them invisibly in the beginning of the task as a simple div. With both methods, the pictures are still initiated a second, sometimes even a third time. Weirdly, this problem does not occur in Firefox, but only in Chrome! Is there an issue with loading them in html versus javascript? I simply can't come up with an explanation. Any suggestions appreciated!

Thanks,
Sophie

**Reply by vijaymarupudi** [reply]
I want to just point out that the server also plays a role in this sometimes. jsPsych just downloads the image, it doesn't really save it anywhere. It relies on the fact that the browser cache will be used to load the image the second time. However if the server sends a [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) header of a certain kind, then it will not cache responses. I would also check the headers your server is sending. (Looks like it's JATOS, but I am not acquainted with it.) 

This link will be useful: <https://www.websparrow.org/misc/how-to-view-http-headers-in-google-chrome>

**Reply by SophieFromm** [reply]
Thanks for the ideas! JATOS should not be a problem here because this is used in the same manner in both browsers on my computer.

I am very new to coding web-based tasks - can I actually set aspects of Cache-Control in my script? When I click quickly through the instructions, the pictures are not loaded again, in contrast to when I wait for longer. In Chrome, "jquery.min.js:3" is initiating those pictures for the second time when a certain time passed after they were loaded the initially. From what could that result, indeed from different browser settings for the Cache-Control?

**Reply by jodeleeuw** [reply]
> can I actually set aspects of Cache-Control in my script

No, this is controlled by the server, not JavaScript. It's my understanding that browsers also have their own internal rules for caching. For example, it looks like Chrome won't cache any pages with faulty security certificates: https://stackoverflow.com/a/35328096/3726673

Is your experiment loading over `https://`?

**Reply by vijaymarupudi** [reply]
The fact that both browsers act differently does not mean that the headers are not the issue. In case no header is provided, browsers are free to optimize sometimes, and they might do this in different ways.

Although in this case, I think one issue might be that Javascript and CSS use different caches, and the code is setting a css property to load the image. I would try to render an `<img>` tag instead if possible.

**Reply by SophieFromm** [reply]
No, it is `http://` and I get the following error (which I stupidyl ignored before):

> "DevTools failed to load SourceMap: Could not load content for chrome-extension://gighmmpiobklfepjocnamgkkbiglidom/include.postload.js.map: HTTP error: status code 404, net::ERR_UNKNOWN_URL_SCHEME"

If caching does not work in Chrome in those cases, we might have gotten to the core of the problem ...

**Reply by jodeleeuw** [reply]
I think `http://` is actually good in this case. The example I linked to indicated that if you were trying to load over `https://` and the security certificate was bad then it wouldn't cache. But loading over `http://` shouldn't cause that issue. 

I strongly suspect that there's some kind of setting in Chrome that's to blame here. One thing that would be useful to know is if you load the page in Chrome on a different computer do you get the same issue? Is it any Chrome browser, or just the one on a particular machine?

**Reply by vijaymarupudi** [reply]
That error is not from jsPsych, it's from a browser extension you have. I think it's safe to ignore, but feel free to disable extensions and test.

**Reply by SophieFromm** [reply]
I figured that it indeed works with `<img>` tags instead of changing css, which leaves me with a good amount of work to do. I am not 100% sure about the role different browsers play though. The behavior is somewhat inconsistent. Different computers (all Windows so far) did not play a role and http vs. https did not make a difference either. In case I find out more details I will post a note for people struggling with similar problems. Thanks for the suggestions!

**Reply by vijaymarupudi** [reply]
The caching mechanism will probably have to be made more browser independent. However, the `<img>` technique should have helped for most browsers. Do you notice an inconsistency in that case?
